mirgecom/diffusion.py

```diff
@@ -49,24 +53,102 @@ def _u_flux(discr, alpha, q_tpair):
     return discr.project(q_tpair.dd, "all_faces", flux_weak)
 
 
-def diffusion_operator(discr, alpha, u):
+class DiffusionBoundary(metaclass=abc.ABCMeta):
+    """
+    Diffusion boundary base class.
+
+    .. automethod:: get_q_flux
+    .. automethod:: get_u_flux
+    """
+
+    @abc.abstractmethod
+    def get_q_flux(self, discr, alpha, dd, u):
+        """Compute the flux for *q* on the boundary corresponding to *dd*."""
+        pass
+
+    def get_u_flux(self, discr, alpha, dd, q):
+        """Compute the flux for *u* on the boundary corresponding to *dd*."""
+        pass
+
+
+class DirichletDiffusionBoundary(DiffusionBoundary):
+    """
+    Dirichlet boundary condition for the diffusion operator.
+
+    .. automethod:: __init__
+    """
+
+    def __init__(self, value=0.):
+        """
+        Initialize the boundary condition.
+
+        Parameters
+        ----------
+        value: float or meshmode.dof_array.DOFArray
+            the value(s) along the boundary
+        """
+        self.value = value
+
+    def get_q_flux(self, discr, alpha, dd, u):  # noqa: D102
+        dir_u = discr.project("vol", dd, u)
+        u_tpair = TracePair(dd, interior=dir_u, exterior=2.*self.value-dir_u)
+        return _q_flux(discr, alpha, u_tpair)
+
+    def get_u_flux(self, discr, alpha, dd, q):  # noqa: D102
+        dir_q = discr.project("vol", dd, q)
+        q_tpair = TracePair(dd, interior=dir_q, exterior=dir_q)
+        return _u_flux(discr, alpha, q_tpair)
+
+
+class NeumannDiffusionBoundary(DiffusionBoundary):
+    """
+    Neumann boundary condition for the diffusion operator.
+
+    .. automethod:: __init__
+    """
+
+    def __init__(self, value=0.):
+        """
+        Initialize the boundary condition.
+
+        Parameters
+        ----------
+        value: float or meshmode.dof_array.DOFArray
+            the value(s) along the boundary
+        """
+        self.value = value
+
+    def get_q_flux(self, discr, alpha, dd, u):  # noqa: D102
+        dir_u = discr.project("vol", dd, u)
```

Comment:
`dir` sseems like a misnomer... (everywhere in Neumann)

Reply:
Ohhhhh... that's what `dir_` means. 🙂 I had been wondering. Fixed.